service/src/main/java/eu/europa/ec/fisheries/uvms/activity/service/dto/fareport/details/VesselDetailsDTO.java

```diff
@@ -124,48 +127,41 @@ public void setStorageDto(StorageDto storageDto) {
         this.storageDto = storageDto;
     }
 
-    public boolean hasEmptyIdentifierValues() {
-
-        if (!CollectionUtils.isEmpty(vesselIdentifiers)) {
-            for (IdentifierDto identifier : vesselIdentifiers) {
-                if (isEmpty(identifier.getFaIdentifierId())) {
-                    return true;
-                }
-            }
+    public void enrichVesselIdentifiersFromAsset(AssetDTO assetDTO) {
+        if (vesselIdentifiers == null) {
+            vesselIdentifiers = new HashSet<>();
+        }
+        if (assetDTO != null) {
+            vesselIdentifiers = addIdIfMissingAndValueProvided(vesselIdentifiers, CFR, assetDTO.getCfr());
+            vesselIdentifiers = addIdIfMissingAndValueProvided(vesselIdentifiers, EXT_MARK, assetDTO.getExternalMarking());
+            vesselIdentifiers = addIdIfMissingAndValueProvided(vesselIdentifiers, GFCM, assetDTO.getGfcm());
+            vesselIdentifiers = addIdIfMissingAndValueProvided(vesselIdentifiers, ICCAT, assetDTO.getIccat());
+            vesselIdentifiers = addIdIfMissingAndValueProvided(vesselIdentifiers, IRCS, assetDTO.getIrcs());
+            vesselIdentifiers = addIdIfMissingAndValueProvided(vesselIdentifiers, UVI, assetDTO.getUvi());
         }
-        return false;
     }
 
-    public void enrichIdentifiers(Asset asset) {
-        if(asset!=null) {
-            String ircs = asset.getIrcs();
-            String cfr = asset.getCfr();
-            String externalMarking = asset.getExternalMarking();
-
-            for (AssetIdentifierDto identifier : vesselIdentifiers) {
-
-                if (isEmpty(identifier.getFaIdentifierId())) {
-
-                    VesselIdentifierSchemeIdEnum identifierSchemeId = identifier.getIdentifierSchemeId();
-
-                    if (CFR.equals(identifierSchemeId) && !isEmpty(cfr)) {
-                        setIdentifier(identifier, cfr);
-                    } else if (EXT_MARK.equals(identifierSchemeId) && !isEmpty(externalMarking)) {
-                        setIdentifier(identifier, externalMarking);
-                    } else if (IRCS.equals(identifierSchemeId) && !isEmpty(ircs)) {
-                        setIdentifier(identifier, ircs);
-                    }
-                }
+    private Set<AssetIdentifierDto> addIdIfMissingAndValueProvided(Set<AssetIdentifierDto> vesselIdentifiers, VesselIdentifierSchemeIdEnum idScheme, String idValue) {
+        AssetIdentifierDto assetIdentifier = null;
+        final Iterator<AssetIdentifierDto> assetIdIterator = vesselIdentifiers.iterator();
+        while (assetIdIterator.hasNext() && assetIdentifier == null) {
+            AssetIdentifierDto tempAssetIdentifierDto = assetIdIterator.next();
+            if (tempAssetIdentifierDto.getIdentifierSchemeId() == idScheme) {
+                assetIdentifier = tempAssetIdentifierDto;
```

Comment:
Can we "break;" after this?

Reply:
We could but the && assetIdentifier == null also prevents from any additional laps in the loop. But I modified the code just below a bit, should be easier to read now